# Programme for International Student Assessment | [PISA](https://www.oecd.org/pisa/) <br>2012 Survey Results Exploration

## by Yousef Abdrabo<br>

# Preliminary Wrangling

> ### Introduction
> The Programme for International Student Assessment is a worldwide study by the Organisation for Economic Co-operation and Development ([OECD](https://www.oecd.org/)) intended to evaluate educational systems by measuring 15-year-old school pupils' scholastic performance on mathematics, science, and reading in form of a survey, this exploration makes use of the dataset of the survey's results to gather and answer some related questions.

### Preparing the Environment

In [1]:
# import all packages and set plots to be embedded inline
import os
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

%matplotlib inline

### Loading Dataset and Obtaining Basic Info

In [2]:
ls

Images/             pisa2012_Wrangled.csv*           README.md*
output_toggle.tpl*  PISA_Results_Exploration.ipynb*  slide_deck.ipynb*
pisa2012.csv*       PISA_Results_Wrangling.ipynb*    slide_deck.slides.html*


In [3]:
df = pd.read_csv('pisa2012.csv', encoding='latin-1', low_memory=False)

In [4]:
size = os.path.getsize('./pisa2012.csv')
print(round(size/int(1e9),1), 'Gigabytes')
del size

2.8 Gigabytes


In [5]:
df.shape

(485490, 636)

## Dataset Introductory Detailed Description
### Dataset Structure

> - **Data Form:** Tabular
>   
> - **Data Container:** CSV file
>   
> - **Size:** 2.8 Gigabytes
>   
> - **Columns:** 636
>   
> - **Rows:** 485490
>   

### Main Points of Interest in The Dataset:

> **1.** Student's Diversity in Everyday-Experieneces<br>
> **2.** Differences in different students' subjects of focus<br>
> **3.** Student's family existance conditon<br>
> **4.** Student's Integration and Satisfaction towards Education<br>
> **5.** Age of starting education<br>
> **6.** Student encouraged towards critical thinking or not<br>
> **7.** Student teaching himself freely with his preffered techniques<br>
> **8.** Student-teacher relation<br>
> **9.** Student learning maths interactively and visually<br>
> **10.** Student language<br>

<hr>

### Features in The Dataset Supporting the Investigation Into The Above Points of Interest:

#### 1. Student's Diversity in Everyday-Experieneces:

> - Student's distance between multiple experienced cultures (CULTDIST)
> - Student plays chess? (column ST49Q06)
> - Student familiar with computer programming? (column ST49Q07)
> - Student has game console? (column IC01Q05)
> - Student has internet connection (column IC01Q04)
> - Student plays single player games out of school? (IC08Q01)
> - Student uses email out of school? (IC08Q03)
> - Student chats online? (IC08Q04)
> - Student on social networks? (IC08Q05)
> - Student browses internet for fun? (IC08Q06)
> - Student reads news? (IC08Q07)
> - Student obtains practical information from the internet? (IC08Q08)
> - Student streams music? (IC08Q09)
> - Student uploads content? (IC08Q11)
> - Student classifies as using ICT for entertainment? (ENTUSE)
> - Student Student limits use of computers as a tool for only schooling? (ICTATTNEG)
> - Student has ICT availability at home? (ICTHOME)
> - Student owns a computer? (ST26Q04)

#### 2. Differences in Different Students' Subjects of Focus

> - Student age (AGE)
> - Student gender (ST04Q01)
> - Student into maths or language courses after school? (ST48Q01)
> - Student into maths or science courses after school? (ST48Q02)
> - Student studies harder in maths vs language? (ST48Q03)
> - Student takes more maths or science classes? (ST48Q04)
> - Student into career of maths vs science? (ST48Q05)
> - Student open for problem solving? (OPENPS)

#### 3. Student's Family Existance Conditon

> - Student mother at home? (ST11Q01)
> - Student father at home? (ST11Q02)
> - Student brothers at home? (ST11Q03)
> - Student sisters at home? (ST11Q04)
> - Student grandparents at home? (ST11Q05)

#### 4. Student's Integration and Satisfaction towards Education

> - Student claims performing poorly regardless? (ST43Q06)
> - Student claims teacher not explaining well? (ST44Q03)
> - Student claims material is too hard? (ST44Q05)
> - Student claims teacher didn't get students interested? (ST44Q07)
> - Student pays attention in classes? (ST46Q06)
> - Student claims teacher provides extra help? (ST77Q02)
> - Student claims teacher provides help in general? (ST77Q04)

#### 5. Age of starting education

> - Student age at Grade ISCED 1 (primary education) (ST06Q01)

#### 6. Student Encouraged Towards Critical Thinking or Not

> - Student's teacher encourages expresssing opinions? (ST77Q06)
> - Student's teacher encourages reasoning? (ST79Q02)
> - Student's teacher gives problems requiring thinking? (ST80Q04)
> - Studen't teacher asks to use own procedures? (ST80Q05)
> - Studen't teacher presents problems with no abvious solutions (ST80Q06)

#### 7. Student Teaches Himself Freely with His Preffered Techniques

> - Out-of-school computer study time (ST57Q06)

#### 8. Student-teacher Relation

> - Student get along with teachers (ST86Q01)
> - Student's teacher is interested (ST86Q02)
> - Student's teacher listens to questions (ST86Q03)
> - Student's teacher gives help (ST86Q04)
> - Student's teacher treats students fair (ST86Q05)

#### 9. Student Learning Maths Interactively and Visually

> - Student draws graphs? (IC11Q01)
> - Student practices? (IC11Q02)
> - Student experiences geometric figures? (IC11Q03)
> - Student experiences spreadsheets? (IC11Q04)
> - Student experiences histograms? (IC11Q06)
> - Student changes in graphs? (IC11Q07)

#### 10. Student Language

> - Student's language at home (LANGN)
<hr>

### Features Used Only to Assist Wrangling

#### To Help Specify Duplicates and Uniques

> - Student Identification Code (STIDSTD)
> - School Identification Code (SCHOOLID)

## Wrangling Final Considerations
### From the Above Detailed Description
* **The columns in the dataset to be made use of in the exploration are the following:**<br>"CULTDIST", "ST49Q06", "ST49Q07", "IC01Q05", "IC01Q04", "IC08Q01", "IC08Q02", "IC08Q03", "IC08Q04", "IC08Q05", "IC08Q06", "IC08Q07", "IC08Q08", "IC08Q09", "IC08Q11", "ENTUSE", "ICTATTNEG", "ICTHOME", "ST26Q04", "AGE", "ST04Q01", "ST48Q01", "ST48Q02", "ST48Q03", "ST48Q04", "ST48Q05", "OPENPS", "ST11Q01", "ST11Q02", "ST11Q03", "ST11Q04", "ST11Q05", "ST43Q06", "ST44Q03", "ST44Q05", "ST44Q07", "ST46Q06", "ST77Q02", "ST77Q04", "ST06Q01", "ST77Q06", "ST79Q02", "ST80Q04", "ST80Q05", "ST80Q06", "ST57Q06", "ST86Q01", "ST86Q02", "ST86Q03", "ST86Q04", "ST86Q05", "IC11Q01", "IC11Q02", "IC11Q03", "IC11Q04", "IC11Q06", "IC11Q07", "LANGN", "SCHOOLID", "STIDSTD" **and so, any other column will be removed to save resources.**
* Even the above columns themselves will be **reduced** and **used to get insights** at the same time by using every related group of columns telling very similar things to construct new unified column speaking for them (e.g. *Student use of ICT for entertainment* and *Student ownership of a game console* etc.), this will save computer resources and make the data more organized without sacrificing any of it.
* Remaining columns will be all renamed to have descriptive labels instead of using code-names

### Further Considerations
* Data places with no value/s (NaNs) will be handled
* Any duplicate rows will be removed leaving only unique rows
<hr>

#### Removing Unused Columns

In [6]:
usedColumns = ["CULTDIST", "ST49Q06", "ST49Q07", "IC01Q05", "IC01Q04", "IC08Q01", "IC08Q02", "IC08Q03", "IC08Q04", "IC08Q05", "IC08Q06", "IC08Q07", "IC08Q08", "IC08Q09", "IC08Q11", "ENTUSE", "ICTATTNEG", "ICTHOME", "ST26Q04", "AGE", "ST04Q01", "ST48Q01", "ST48Q02", "ST48Q03", "ST48Q04", "ST48Q05", "OPENPS", "ST11Q01", "ST11Q02", "ST11Q03", "ST11Q04", "ST11Q05", "ST43Q06", "ST44Q03", "ST44Q05", "ST44Q07", "ST46Q06", "ST77Q02", "ST77Q04", "ST13Q01", "ST17Q01", "ST06Q01", "ST77Q06", "ST79Q02", "ST80Q04", "ST80Q05", "ST80Q06", "ST57Q06", "ST86Q01", "ST86Q02", "ST86Q03", "ST86Q04", "ST86Q05", "IC11Q01", "IC11Q02", "IC11Q03", "IC11Q04", "IC11Q06", "IC11Q07", "LANGN", "STIDSTD", "SCHOOLID"]
unusedColumns = list()
for column in df.columns:
    if(column not in usedColumns):
        unusedColumns.append(column)
df.drop(unusedColumns, axis = "columns", inplace=True)
if(len(usedColumns) == len(df.columns)):
    print('Unused columns has been removed successfully')

Unused columns has been removed successfully


#### Removing Duplicates in Rows

In [7]:
print('Shape before removal:', df.shape)
df.drop_duplicates(inplace=True)
print('Shape after removal:', df.shape)

Shape before removal: (485490, 62)
Shape after removal: (485488, 62)


#### Giving Columns Meaningful Names

In [8]:
df.rename(columns={'CULTDIST': 'cultDist',
                   'ST49Q06': 'playsChess',
                   'ST49Q07': 'familairWCode',
                   'IC01Q05': 'hasPS',
                   'IC01Q04': 'hasInternet',
                   'IC08Q01': 'playsSPGames',
                   'IC08Q02': 'playsCOOP',
                   'IC08Q03': 'usesEmail',
                   'IC08Q04': 'chatsOnline',
                   'IC08Q05': 'onFB',
                   'IC08Q06': 'internetFun',
                   'IC08Q07': 'readsNews',
                   'IC08Q08': 'internetLearn',
                   'IC08Q09': 'hasMusic',
                   'IC08Q11': 'uploads',
                   'ENTUSE': 'ICTForFun',
                   'ICTATTNEG': 'UsesPCForEduOnly',
                   'ICTHOME': 'ICTHome',
                   'ST26Q04': 'hasPC',
                   'AGE' : 'age',
                   'ST04Q01': 'sex',
                   'ST48Q01': 'mathsOrLangCourses',
                   'ST48Q02': 'mathsOrSciCourses',
                   'ST48Q03': 'studyMathsOrLang',
                   'ST48Q04': 'studyMathsOrSci',
                   'ST48Q05': 'mathsOrSciJob',
                   'OPENPS': 'problemSolver',
                   'ST11Q01': 'motherAtHome',
                   'ST11Q02': 'fatherAtHome',
                   'ST11Q03': 'brothersAtHome',
                   'ST11Q04': 'sistersAtHome',
                   'ST11Q05': 'granniesAtHomes',
                   'ST43Q06': 'saysDoPoorlyRegardless',
                   'ST44Q03': 'teacherNoExplain',
                   'ST44Q05': 'saysMatsSoHard',
                   'ST44Q07': 'saysTeacherNoMakingStudentInterested',
                   'ST46Q06': 'paysAttInClass',
                   'ST77Q02': 'saysTeacherGivesExtraHelp',
                   'ST77Q04': 'saysTeacherGivesHelp',
                   'ST06Q01': 'ageStartedEdu',
                   'ST77Q06': 'teacherAllowsExpOpinion',
                   'ST79Q02': 'teacherAllowsReasoning',
                   'ST80Q04': 'teacherTrainsThinking',
                   'ST80Q05': 'teacherAllowsOwnProcedures',
                   'ST80Q06': 'teacherAllowsMultipleSolutions',
                   'ST57Q06': 'studyOnPCOutSchoolTime',
                   'ST86Q01': 'copeWithTeacher',
                   'ST86Q02': 'teacherPaysInterest',
                   'ST86Q03': 'teacherListen',
                   'ST86Q04': 'teacherHelps',
                   'ST86Q05': 'teacherIsFair',
                   'IC11Q01': 'drawsGraphs',
                   'IC11Q02': 'practices',
                   'IC11Q03': 'experiencesFigures',
                   'IC11Q04': 'experiencesSpSheets',
                   'IC11Q06': 'experiencesHistograms',
                   'IC11Q07': 'interactsGraphs',
                   'LANGN' : 'langAtHome',
                   'SCHOOLID': 'schoolID',
                   'STIDSTD' : 'studentID'}, inplace=True)

#### Removing Columns With Mostly No Data

In [9]:
rows = df.shape[0]
columnsBefore = df.shape[1]
nanCountPerColumn = df.isna().sum().sort_values(ascending=False)
nanPercentagePerColumn = round((nanCountPerColumn / rows) * 100)
columnsWithNoData = list()
for column in nanPercentagePerColumn.index:
    if(nanPercentagePerColumn[column] > 50):
        columnsWithNoData.append(column)
print('No-data columns to be removed:')
for column in columnsWithNoData:
    print("'"+column+"'")
print("-"*15)
df.drop(columnsWithNoData, axis='columns', inplace=True)
columnsAfter = df.shape[1]
if(columnsBefore - columnsAfter == len(columnsWithNoData)):
    print('Removed successfully.')

No-data columns to be removed:
'cultDist'
---------------
Removed successfully.


### Transforming Text-based Data To Binary Based

#### 'playsChess' Column

In [10]:
# getting an idea about the column data
df['playsChess'].value_counts()

Never or rarely            183262
Sometimes                   75817
Often                       35141
Always or almost always     18351
Name: playsChess, dtype: int64

In [11]:
# transforming data
df['playsChess'] = df['playsChess'].map({'Never or rarely': 0,
                                         'Sometimes': 0,
                                         'Often': 1,
                                        'Always or almost always': 1})
df['playsChess'].value_counts()

0.0    259079
1.0     53492
Name: playsChess, dtype: int64

<hr>

#### 'familairWCode' Column

In [12]:
# getting an idea about the column data
df['familairWCode'].value_counts()

Never or rarely            185257
Sometimes                   68441
Often                       37373
Always or almost always     21354
Name: familairWCode, dtype: int64

In [13]:
# transforming data
df['familairWCode'] = df['familairWCode'].map({'Never or rarely': 0,
                                         'Sometimes': 0,
                                         'Often': 1,
                                        'Always or almost always': 1})
df['familairWCode'].value_counts()

0.0    253698
1.0     58727
Name: familairWCode, dtype: int64

<hr>

#### 'hasPS' Column

In [14]:
# getting an idea about the column data
df['hasPS'].value_counts()

Yes, and I use it          151730
No                          98253
Yes, but I don’t use it     46604
Name: hasPS, dtype: int64

In [15]:
# transforming data
df['hasPS'] = df['hasPS'].map({'No': 0,
                               'Yes, and I use it': 1,
                               'Yes, but I dont use it': 1})
df['hasPS'].value_counts()

1.0    151730
0.0     98253
Name: hasPS, dtype: int64

<hr>

#### 'hasInternet' Column

In [16]:
# getting an idea about the column data
df['hasInternet'].value_counts()

Yes, and I use it          259762
No                          30161
Yes, but I don’t use it      7382
Name: hasInternet, dtype: int64

In [17]:
# transforming data
df['hasInternet'] = df['hasInternet'].map({'No': 0,
                               'Yes, and I use it': 1,
                               'Yes, but I dont use it': 1})
df['hasInternet'].value_counts()

1.0    259762
0.0     30161
Name: hasInternet, dtype: int64

<hr>

#### 'playsSPGames' Column

In [18]:
# getting an idea about the column data
df['playsSPGames'].value_counts()

Never or hardly ever     123799
Once or twice a month     57815
Once or twice a week      57747
Almost every day          33542
Every day                 21220
Name: playsSPGames, dtype: int64

In [19]:
# transforming data
df['playsSPGames'] = df['playsSPGames'].map({'Never or hardly ever': 0,
                               'Once or twice a month': 1,
                               'Once or twice a week': 1,
                                'Almost every day': 1, 
                                            'Every day': 1})
df['playsSPGames'].value_counts()

1.0    170324
0.0    123799
Name: playsSPGames, dtype: int64

<hr>

#### 'playsCOOP' Column

In [20]:
# getting an idea about the column data
df['playsCOOP'].value_counts()

Never or hardly ever     153713
Once or twice a month     42180
Once or twice a week      40127
Almost every day          29563
Every day                 28063
Name: playsCOOP, dtype: int64

In [21]:
# transforming data
df['playsCOOP'] = df['playsCOOP'].map({'Never or hardly ever': 0,
                               'Once or twice a month': 1,
                               'Once or twice a week': 1,
                                'Almost every day': 1, 
                                            'Every day': 1})
df['playsCOOP'].value_counts()

0.0    153713
1.0    139933
Name: playsCOOP, dtype: int64

<hr>

#### 'usesEmail' Column

In [22]:
# getting an idea about the column data
df['usesEmail'].value_counts()

Once or twice a week     75138
Once or twice a month    61905
Never or hardly ever     56927
Almost every day         56510
Every day                42682
Name: usesEmail, dtype: int64

In [23]:
# transforming data
df['usesEmail'] = df['usesEmail'].map({'Never or hardly ever': 0,
                               'Once or twice a month': 1,
                               'Once or twice a week': 1,
                                'Almost every day': 1, 
                                            'Every day': 1})
df['usesEmail'].value_counts()

1.0    236235
0.0     56927
Name: usesEmail, dtype: int64

<hr>

#### 'chatsOnline' Column

In [24]:
# getting an idea about the column data
df['chatsOnline'].value_counts()

Every day                86478
Almost every day         66022
Never or hardly ever     64694
Once or twice a week     45904
Once or twice a month    30151
Name: chatsOnline, dtype: int64

In [25]:
# transforming data
df['chatsOnline'] = df['chatsOnline'].map({'Never or hardly ever': 0,
                               'Once or twice a month': 1,
                               'Once or twice a week': 1,
                                'Almost every day': 1, 
                                            'Every day': 1})
df['chatsOnline'].value_counts()

1.0    228555
0.0     64694
Name: chatsOnline, dtype: int64

<hr>

#### 'onFB' Column

In [26]:
# getting an idea about the column data
df['onFB'].value_counts()

Every day                143204
Almost every day          64272
Never or hardly ever      38051
Once or twice a week      32452
Once or twice a month     15843
Name: onFB, dtype: int64

In [27]:
# transforming data
df['onFB'] = df['onFB'].map({'Never or hardly ever': 0,
                               'Once or twice a month': 1,
                               'Once or twice a week': 1,
                                'Almost every day': 1, 
                                            'Every day': 1})
df['onFB'].value_counts()

1.0    255771
0.0     38051
Name: onFB, dtype: int64

<hr>

#### 'internetFun' Column

In [28]:
# getting an idea about the column data
df['internetFun'].value_counts()

Every day                116715
Almost every day          79288
Once or twice a week      53339
Never or hardly ever      22307
Once or twice a month     22095
Name: internetFun, dtype: int64

In [29]:
# transforming data
df['internetFun'] = df['internetFun'].map({'Never or hardly ever': 0,
                               'Once or twice a month': 1,
                               'Once or twice a week': 1,
                                'Almost every day': 1, 
                                            'Every day': 1})
df['internetFun'].value_counts()

1.0    271437
0.0     22307
Name: internetFun, dtype: int64

<hr>

#### 'readsNews' Column

In [30]:
# getting an idea about the column data
df['readsNews'].value_counts()

Never or hardly ever     68524
Once or twice a week     68439
Almost every day         56580
Once or twice a month    50995
Every day                49032
Name: readsNews, dtype: int64

In [31]:
# transforming data
df['readsNews'] = df['readsNews'].map({'Never or hardly ever': 0,
                               'Once or twice a month': 1,
                               'Once or twice a week': 1,
                                'Almost every day': 1, 
                                            'Every day': 1})
df['readsNews'].value_counts()

1.0    225046
0.0     68524
Name: readsNews, dtype: int64

<hr>

#### 'internetLearn' Column

In [32]:
# getting an idea about the column data
df['internetLearn'].value_counts()

Once or twice a week     87419
Almost every day         60668
Once or twice a month    57380
Never or hardly ever     48274
Every day                39312
Name: internetLearn, dtype: int64

In [33]:
# transforming data
df['internetLearn'] = df['internetLearn'].map({'Never or hardly ever': 0,
                               'Once or twice a month': 1,
                               'Once or twice a week': 1,
                                'Almost every day': 1, 
                                            'Every day': 1})
df['internetLearn'].value_counts()

1.0    244779
0.0     48274
Name: internetLearn, dtype: int64

<hr>

#### 'hasMusic' Column

In [34]:
# getting an idea about the column data
df['hasMusic'].value_counts()

Every day                71417
Once or twice a week     67534
Almost every day         67300
Once or twice a month    45129
Never or hardly ever     42116
Name: hasMusic, dtype: int64

In [35]:
# transforming data
df['hasMusic'] = df['hasMusic'].map({'Never or hardly ever': 0,
                               'Once or twice a month': 1,
                               'Once or twice a week': 1,
                                'Almost every day': 1, 
                                            'Every day': 1})
df['hasMusic'].value_counts()

1.0    251380
0.0     42116
Name: hasMusic, dtype: int64

<hr>

#### 'uploads' Column

In [36]:
# getting an idea about the column data
df['uploads'].value_counts()

Never or hardly ever     144977
Once or twice a month     51538
Once or twice a week      40221
Every day                 28393
Almost every day          28302
Name: uploads, dtype: int64

In [37]:
# transforming data
df['uploads'] = df['uploads'].map({'Never or hardly ever': 0,
                               'Once or twice a month': 1,
                               'Once or twice a week': 1,
                                'Almost every day': 1, 
                                            'Every day': 1})
df['uploads'].value_counts()

1.0    148454
0.0    144977
Name: uploads, dtype: int64

<hr>

#### 'hasPC' Column

In [38]:
# getting an idea about the column data
df['hasPC'].value_counts()

Yes    411040
No      62837
Name: hasPC, dtype: int64

In [39]:
# transforming data
df['hasPC'] = df['hasPC'].map({'Yes': 1,
                               'No': 0,
                               })
df['hasPC'].value_counts()

1.0    411040
0.0     62837
Name: hasPC, dtype: int64

<hr>

#### 'mathsOrLangCourses' Column

In [40]:
# getting an idea about the column data
df['mathsOrLangCourses'].value_counts()

Courses after school Math             171352
Courses after school Test Language    123058
Name: mathsOrLangCourses, dtype: int64

In [41]:
# transforming data
df['mathsOrLangCourses'] = df['mathsOrLangCourses'].map({'Courses after school Math': 'Maths',
                               'Courses after school Test Language': 'Language',
                               })
df['mathsOrLangCourses'].value_counts()

Maths       171352
Language    123058
Name: mathsOrLangCourses, dtype: int64

<hr>

#### 'mathsOrSciCourses' Column

In [42]:
# getting an idea about the column data
df['mathsOrSciCourses'].value_counts()

Major in college Science    167095
Major in college Math       122732
Name: mathsOrSciCourses, dtype: int64

In [43]:
# transforming data
df['mathsOrSciCourses'] = df['mathsOrSciCourses'].map({'Major in college Science': 'Science',
                               'Major in college Math': 'Math',
                               })
df['mathsOrSciCourses'].value_counts()

Science    167095
Math       122732
Name: mathsOrSciCourses, dtype: int64

<hr>

#### 'studyMathsOrLang' Column

In [44]:
# getting an idea about the column data
df['studyMathsOrLang'].value_counts()

Study harder Math             178180
Study harder Test Language    120299
Name: studyMathsOrLang, dtype: int64

In [45]:
# transforming data
df['studyMathsOrLang'] = df['studyMathsOrLang'].map({'Study harder Math': 'Math',
                               'Study harder Test Language': 'Language',
                               })
df['studyMathsOrLang'].value_counts()

Math        178180
Language    120299
Name: studyMathsOrLang, dtype: int64

<hr>

#### 'studyMathsOrSci' Column

In [46]:
# getting an idea about the column data
df['studyMathsOrSci'].value_counts()

Maximum classes Math       134002
Maximum classes Science    133714
Name: studyMathsOrSci, dtype: int64

In [47]:
# transforming data
df['studyMathsOrSci'] = df['studyMathsOrSci'].map({'Maximum classes Math': 'Math',
                               'Maximum classes Science': 'Science',
                               })
df['studyMathsOrSci'].value_counts()

Math       134002
Science    133714
Name: studyMathsOrSci, dtype: int64

<hr>

#### 'mathsOrSciJob' Column

In [48]:
# getting an idea about the column data
df['mathsOrSciJob'].value_counts()

Pursuing a career Science    161633
Pursuing a career Math       126359
Name: mathsOrSciJob, dtype: int64

In [49]:
# transforming data
df['mathsOrSciJob'] = df['mathsOrSciJob'].map({'Pursuing a career Science': 'Science',
                               'Pursuing a career Math': 'Math',
                               })
df['mathsOrSciJob'].value_counts()

Science    161633
Math       126359
Name: mathsOrSciJob, dtype: int64

<hr>

#### 'motherAtHome' Column

In [50]:
# getting an idea about the column data
df['motherAtHome'].value_counts()

Yes    433007
No      27552
Name: motherAtHome, dtype: int64

In [51]:
# transforming data
df['motherAtHome'] = df['motherAtHome'].map({'Yes': 1,
                               'No': 0,
                               })
df['motherAtHome'].value_counts()

1.0    433007
0.0     27552
Name: motherAtHome, dtype: int64

<hr>

#### 'fatherAtHome' Column

In [52]:
# getting an idea about the column data
df['fatherAtHome'].value_counts()

Yes    372161
No      68875
Name: fatherAtHome, dtype: int64

In [53]:
# transforming data
df['fatherAtHome'] = df['fatherAtHome'].map({'Yes': 1,
                               'No': 0,
                               })
df['fatherAtHome'].value_counts()

1.0    372161
0.0     68875
Name: fatherAtHome, dtype: int64

<hr>

#### 'brothersAtHome' Column

In [54]:
# getting an idea about the column data
df['brothersAtHome'].value_counts()

Yes    245262
No     154814
Name: brothersAtHome, dtype: int64

In [55]:
# transforming data
df['brothersAtHome'] = df['brothersAtHome'].map({'Yes': 1,
                               'No': 0,
                               })
df['brothersAtHome'].value_counts()

1.0    245262
0.0    154814
Name: brothersAtHome, dtype: int64

<hr>

#### 'sistersAtHome' Column

In [56]:
# getting an idea about the column data
df['sistersAtHome'].value_counts()

Yes    223061
No     167707
Name: sistersAtHome, dtype: int64

In [57]:
# transforming data
df['sistersAtHome'] = df['sistersAtHome'].map({'Yes': 1,
                               'No': 0,
                               })
df['sistersAtHome'].value_counts()

1.0    223061
0.0    167707
Name: sistersAtHome, dtype: int64

<hr>

#### 'granniesAtHomes' Column

In [58]:
# getting an idea about the column data
df['granniesAtHomes'].value_counts()

No     275636
Yes     72544
Name: granniesAtHomes, dtype: int64

In [59]:
# transforming data
df['granniesAtHomes'] = df['granniesAtHomes'].map({'Yes': 1,
                               'No': 0,
                               })
df['granniesAtHomes'].value_counts()

0.0    275636
1.0     72544
Name: granniesAtHomes, dtype: int64

<hr>

#### 'saysDoPoorlyRegardless' Column

In [60]:
# getting an idea about the column data
df['saysDoPoorlyRegardless'].value_counts()

Disagree             127111
Strongly disagree     94150
Agree                 63524
Strongly agree        28685
Name: saysDoPoorlyRegardless, dtype: int64

In [61]:
# transforming data
df['saysDoPoorlyRegardless'] = df['saysDoPoorlyRegardless'].map({'Agree': 1,
                               'Strongly disagree': 0,
                                'Disagree': 0,
                                'Strongly agree': 1
                               })
df['saysDoPoorlyRegardless'].value_counts()

0.0    221261
1.0     92209
Name: saysDoPoorlyRegardless, dtype: int64

<hr>

#### 'teacherNoExplain' Column

In [62]:
# getting an idea about the column data
df['teacherNoExplain'].value_counts()

Slightly likely      109357
Likely               104832
Not at all likely     59997
Very   Likely         39219
Name: teacherNoExplain, dtype: int64

In [63]:
# transforming data
df['teacherNoExplain'] = df['teacherNoExplain'].map({'Slightly likely': 1,
                               'Likely': 1,
                                    'Not at all likely':0,
                                        'Very   Likely':1
                               })
df['teacherNoExplain'].value_counts()

1.0    253408
0.0     59997
Name: teacherNoExplain, dtype: int64

<hr>

#### 'saysMatsSoHard' Column

In [64]:
# getting an idea about the column data
df['saysMatsSoHard'].value_counts()

Likely               139516
Very   Likely         80794
Slightly likely       67979
Not at all likely     24707
Name: saysMatsSoHard, dtype: int64

In [65]:
# transforming data
df['saysMatsSoHard'] = df['saysMatsSoHard'].map({'Slightly likely': 1,
                               'Likely': 1,
                                    'Not at all likely':0,
                                        'Very   Likely':1})
df['saysMatsSoHard'].value_counts()

1.0    288289
0.0     24707
Name: saysMatsSoHard, dtype: int64

<hr>

#### 'saysTeacherNoMakingStudentInterested' Column

In [66]:
# getting an idea about the column data
df['saysTeacherNoMakingStudentInterested'].value_counts()

Likely               101953
Slightly likely       99750
Very   Likely         55967
Not at all likely     55300
Name: saysTeacherNoMakingStudentInterested, dtype: int64

In [67]:
# transforming data
df['saysTeacherNoMakingStudentInterested'] = df['saysTeacherNoMakingStudentInterested'].map({'Slightly likely': 1,
                               'Likely': 1,
                                    'Not at all likely':0,
                                        'Very   Likely':1})
df['saysTeacherNoMakingStudentInterested'].value_counts()

1.0    257670
0.0     55300
Name: saysTeacherNoMakingStudentInterested, dtype: int64

<hr>

#### 'paysAttInClass' Column

In [68]:
# getting an idea about the column data
df['paysAttInClass'].value_counts()

Agree                171776
Strongly agree        82713
Disagree              48412
Strongly disagree      9999
Name: paysAttInClass, dtype: int64

In [69]:
# transforming data
df['paysAttInClass'] = df['paysAttInClass'].map({'Agree': 1,
                               'Strongly agree': 1,
                                    'Disagree':0,
                                        'Strongly disagree':0})
df['paysAttInClass'].value_counts()

1.0    254489
0.0     58411
Name: paysAttInClass, dtype: int64

<hr>

#### 'saysTeacherGivesExtraHelp' Column

In [70]:
# getting an idea about the column data
df['saysTeacherGivesExtraHelp'].value_counts()

Every Lesson            136185
Most Lessons             94283
Some Lessons             64852
Never or Hardly Ever     19593
Name: saysTeacherGivesExtraHelp, dtype: int64

In [71]:
# transforming data
df['saysTeacherGivesExtraHelp'] = df['saysTeacherGivesExtraHelp'].map({'Every Lesson': 1,
                               'Most Lessons': 1,
                                    'Some Lessons':0,
                                        'Never or Hardly Ever':0})
df['saysTeacherGivesExtraHelp'].value_counts()

1.0    230468
0.0     84445
Name: saysTeacherGivesExtraHelp, dtype: int64

<hr>

#### 'saysTeacherGivesHelp' Column

In [72]:
# getting an idea about the column data
df['saysTeacherGivesHelp'].value_counts()

Every Lesson            153025
Most Lessons             89992
Some Lessons             54126
Never or Hardly Ever     17225
Name: saysTeacherGivesHelp, dtype: int64

In [73]:
# transforming data
df['saysTeacherGivesHelp'] = df['saysTeacherGivesHelp'].map({'Every Lesson': 1,
                               'Most Lessons': 1,
                                    'Some Lessons':0,
                                        'Never or Hardly Ever':0})
df['saysTeacherGivesHelp'].value_counts()

1.0    243017
0.0     71351
Name: saysTeacherGivesHelp, dtype: int64

<hr>

#### 'teacherAllowsExpOpinion' Column

In [74]:
# getting an idea about the column data
df['teacherAllowsExpOpinion'].value_counts()

Every Lesson            133732
Most Lessons             89813
Some Lessons             63867
Never or Hardly Ever     27395
Name: teacherAllowsExpOpinion, dtype: int64

In [75]:
# transforming data
df['teacherAllowsExpOpinion'] = df['teacherAllowsExpOpinion'].map({'Every Lesson': 1,
                               'Most Lessons': 1,
                                    'Some Lessons':0,
                                        'Never or Hardly Ever':0})
df['teacherAllowsExpOpinion'].value_counts()

1.0    223545
0.0     91262
Name: teacherAllowsExpOpinion, dtype: int64

<hr>

#### 'teacherAllowsReasoning' Column

In [76]:
# getting an idea about the column data
df['teacherAllowsReasoning'].value_counts()

Most Lessons            104082
Some Lessons             92573
Every Lesson             73557
Never or Hardly Ever     44116
Name: teacherAllowsReasoning, dtype: int64

In [77]:
# transforming data
df['teacherAllowsReasoning'] = df['teacherAllowsReasoning'].map({'Every Lesson': 1,
                               'Most Lessons': 1,
                                    'Some Lessons':0,
                                        'Never or Hardly Ever':0})
df['teacherAllowsReasoning'].value_counts()

1.0    177639
0.0    136689
Name: teacherAllowsReasoning, dtype: int64

<hr>

#### 'teacherTrainsThinking' Column

In [78]:
# getting an idea about the column data
df['teacherTrainsThinking'].value_counts()

Often                      115669
Sometimes                  114908
Always or almost always     56817
Never or rarely             26127
Name: teacherTrainsThinking, dtype: int64

In [79]:
# transforming data
df['teacherTrainsThinking'] = df['teacherTrainsThinking'].map({'Never or rarely': 0,
                               'Often': 1,
                                    'Sometimes':0,
                                        'Always or almost always':1})
df['teacherTrainsThinking'].value_counts()

1.0    172486
0.0    141035
Name: teacherTrainsThinking, dtype: int64

<hr>

#### 'teacherAllowsOwnProcedures' Column

In [80]:
# getting an idea about the column data
df['teacherAllowsOwnProcedures'].value_counts()

Sometimes                  113172
Often                       87900
Never or rarely             61844
Always or almost always     49677
Name: teacherAllowsOwnProcedures, dtype: int64

In [81]:
# transforming data
df['teacherAllowsOwnProcedures'] = df['teacherAllowsOwnProcedures'].map({'Never or rarely': 0,
                               'Often': 1,
                                    'Sometimes':0,
                                        'Always or almost always':1})
df['teacherAllowsOwnProcedures'].value_counts()

0.0    175016
1.0    137577
Name: teacherAllowsOwnProcedures, dtype: int64

<hr>

#### 'teacherAllowsMultipleSolutions' Column

In [82]:
# getting an idea about the column data
df['teacherAllowsMultipleSolutions'].value_counts()

Sometimes                  112906
Often                       95443
Never or rarely             55758
Always or almost always     48383
Name: teacherAllowsMultipleSolutions, dtype: int64

In [83]:
# transforming data
df['teacherAllowsMultipleSolutions'] = df['teacherAllowsMultipleSolutions'].map({'Never or rarely': 0,
                               'Often': 1,
                                    'Sometimes':0,
                                        'Always or almost always':1})
df['teacherAllowsMultipleSolutions'].value_counts()

0.0    168664
1.0    143826
Name: teacherAllowsMultipleSolutions, dtype: int64

<hr>

#### 'copeWithTeacher' Column

In [84]:
# getting an idea about the column data
df['copeWithTeacher'].value_counts()

Agree                180918
Strongly agree        81459
Disagree              43810
Strongly disagree      7036
Name: copeWithTeacher, dtype: int64

In [85]:
# transforming data
df['copeWithTeacher'] = df['copeWithTeacher'].map({'Agree': 1,
                               'Strongly agree': 1,
                                    'Disagree':0,
                                        'Strongly disagree':0})
df['copeWithTeacher'].value_counts()

1.0    262377
0.0     50846
Name: copeWithTeacher, dtype: int64

<hr>

#### 'teacherPaysInterest' Column

In [86]:
# getting an idea about the column data
df['teacherPaysInterest'].value_counts()

Agree                175300
Strongly agree        76449
Disagree              51359
Strongly disagree      9483
Name: teacherPaysInterest, dtype: int64

In [87]:
# transforming data
df['teacherPaysInterest'] = df['teacherPaysInterest'].map({'Agree': 1,
                               'Strongly agree': 1,
                                    'Disagree':0,
                                        'Strongly disagree':0})
df['teacherPaysInterest'].value_counts()

1.0    251749
0.0     60842
Name: teacherPaysInterest, dtype: int64

<hr>

#### 'teacherListen' Column

In [88]:
# getting an idea about the column data
df['teacherListen'].value_counts()

Agree                168284
Strongly agree        69420
Disagree              62855
Strongly disagree     11629
Name: teacherListen, dtype: int64

In [89]:
# transforming data
df['teacherListen'] = df['teacherListen'].map({'Agree': 1,
                               'Strongly agree': 1,
                                    'Disagree':0,
                                        'Strongly disagree':0})
df['teacherListen'].value_counts()

1.0    237704
0.0     74484
Name: teacherListen, dtype: int64

<hr>

#### 'teacherHelps' Column

In [90]:
# getting an idea about the column data
df['teacherHelps'].value_counts()

Agree                173260
Strongly agree        86203
Disagree              42840
Strongly disagree      9991
Name: teacherHelps, dtype: int64

In [91]:
# transforming data
df['teacherHelps'] = df['teacherHelps'].map({'Agree': 1,
                               'Strongly agree': 1,
                                    'Disagree':0,
                                        'Strongly disagree':0})
df['teacherHelps'].value_counts()

1.0    259463
0.0     52831
Name: teacherHelps, dtype: int64

<hr>

#### 'teacherIsFair' Column

In [92]:
# getting an idea about the column data
df['teacherIsFair'].value_counts()

Agree                174259
Strongly agree        84213
Disagree              40843
Strongly disagree     12589
Name: teacherIsFair, dtype: int64

In [93]:
# transforming data
df['teacherIsFair'] = df['teacherIsFair'].map({'Agree': 1,
                               'Strongly agree': 1,
                                    'Disagree':0,
                                        'Strongly disagree':0})
df['teacherIsFair'].value_counts()

1.0    258472
0.0     53432
Name: teacherIsFair, dtype: int64

<hr>

#### 'drawsGraphs' Column

In [94]:
# getting an idea about the column data
df['drawsGraphs'].value_counts()

No                                              198909
Yes, students did this                           49745
Yes, but only the teacher demonstrated this      41240
Name: drawsGraphs, dtype: int64

In [95]:
# transforming data
df['drawsGraphs'] = df['drawsGraphs'].map({'No': 0,
                               'Yes, students did this': 1,
                                    'Yes, but only the teacher demonstrated this':0
                                          })
df['drawsGraphs'].value_counts()

0.0    198909
1.0     49745
Name: drawsGraphs, dtype: int64

<hr>

#### 'practices' Column

In [96]:
# getting an idea about the column data
df['practices'].value_counts()

No                                              212111
Yes, students did this                           44707
Yes, but only the teacher demonstrated this      32609
Name: practices, dtype: int64

In [97]:
# transforming data
df['practices'] = df['practices'].map({'No': 0,
                               'Yes, students did this': 1,
                                    'Yes, but only the teacher demonstrated this':0
                                          })
df['practices'].value_counts()

0.0    212111
1.0     44707
Name: practices, dtype: int64

<hr>

#### 'experiencesFigures' Column

In [98]:
# getting an idea about the column data
df['experiencesFigures'].value_counts()

No                                              203795
Yes, students did this                           44102
Yes, but only the teacher demonstrated this      40971
Name: experiencesFigures, dtype: int64

In [99]:
# transforming data
df['experiencesFigures'] = df['experiencesFigures'].map({'No': 0,
                               'Yes, students did this': 1,
                                    'Yes, but only the teacher demonstrated this':0
                                          })
df['experiencesFigures'].value_counts()

0.0    203795
1.0     44102
Name: experiencesFigures, dtype: int64

<hr>

#### 'experiencesSpSheets' Column

In [100]:
# getting an idea about the column data
df['experiencesSpSheets'].value_counts()

No                                              194896
Yes, students did this                           57750
Yes, but only the teacher demonstrated this      36439
Name: experiencesSpSheets, dtype: int64

In [101]:
# transforming data
df['experiencesSpSheets'] = df['experiencesSpSheets'].map({'No': 0,
                               'Yes, students did this': 1,
                                    'Yes, but only the teacher demonstrated this':0
                                          })
df['experiencesSpSheets'].value_counts()

0.0    194896
1.0     57750
Name: experiencesSpSheets, dtype: int64

<hr>

#### 'experiencesHistograms' Column

In [102]:
# getting an idea about the column data
df['experiencesHistograms'].value_counts()

No                                              213321
Yes, students did this                           38592
Yes, but only the teacher demonstrated this      37356
Name: experiencesHistograms, dtype: int64

In [103]:
# transforming data
df['experiencesHistograms'] = df['experiencesHistograms'].map({'No': 0,
                               'Yes, students did this': 1,
                                    'Yes, but only the teacher demonstrated this':0
                                          })
df['experiencesHistograms'].value_counts()

0.0    213321
1.0     38592
Name: experiencesHistograms, dtype: int64

<hr>

#### 'interactsGraphs' Column

In [104]:
# getting an idea about the column data
df['interactsGraphs'].value_counts()

No                                              210077
Yes, but only the teacher demonstrated this      40090
Yes, students did this                           39110
Name: interactsGraphs, dtype: int64

In [105]:
# transforming data
df['interactsGraphs'] = df['interactsGraphs'].map({'No': 0,
                               'Yes, students did this': 1,
                                    'Yes, but only the teacher demonstrated this':0
                                          })
df['interactsGraphs'].value_counts()

0.0    210077
1.0     39110
Name: interactsGraphs, dtype: int64

<p><center>End of transforming data.</center></p>
<hr>

### Unifying Columns Telling Nearly the Same Features of a Student

#### A. Declaring Helping Function/s:

In [106]:
def unifySoSimilarBinaryColumns(dataframe ,soSimilarBinaryColumns):
    """
    - This function takes a list of strings representing labels of any dataframe columns of choice holding binary data that are telling very similar features (e.g. two columns, one telling whether or not a record is associated with being professional, and the other telling whether or not a record is associated with being expert), after this, the function produces and returns a new dataframe column (series) holding the value 1 whenever any of the similar columns' values is equal to 1, or holding the value 0 whenever all the similar columns' values are equal to 0 .
    - All columns should belong to the same dataframe, given by argument 'dataframe'
    """
    #creating a copy of dataframe for temporary usage not to mess with the original dataframe (pass by value approach)
    dataframeCopy = dataframe.copy()
    
    # Initializing the resulting column
    dataForResultingColumn = np.full([ 1, dataframe.shape[0] ], np.nan)
    dataframeCopy['resultingColumn'] = pd.Series(dataForResultingColumn[0]) # the 1d array of the index 0 in 'dataForResultingColumn' is the only 1d array inside the 2d array, so no worries
    
    # Two arrays to hold indecies of dataframe where the value of resulting column should be 1 and another when it should be 0 (respectively)
    recordsWhereValueForResultingColumnShouldBe1 = np.empty((0))
    recordsWhereValueForResultingColumnShouldBe0 = np.empty((0))
    
    # Get indecies where resulting column value should be 1 ...
    # ----------------------------------------------------------
    for column in soSimilarBinaryColumns:
        # get indecies where this particular column value is 1 (which implies where resulting column should also be one)
        indeciesOfThatColBeing1 = dataframeCopy.query("{} == 1".format(column)).index
        # append these indecies to the array holding indecies where the value of resulting column should be 1 ('recordsWhereValueForResultingColumnShouldBe1')
        recordsWhereValueForResultingColumnShouldBe1 = np.append(recordsWhereValueForResultingColumnShouldBe1, indeciesOfThatColBeing1)
    #remove same indecies appended more than one time (if exist)
    recordsWhereValueForResultingColumnShouldBe1 = np.unique(recordsWhereValueForResultingColumnShouldBe1)
    
    # Get indecies where resulting column value should be 0 ...
    # ----------------------------------------------------------
    for column in soSimilarBinaryColumns:
        # get indecies where this particular column value is 0
        indeciesOfThatColBeing0 = dataframeCopy.query("{} == 0".format(column)).index
        # append these indecies to the array holding indecies where the value of resulting column should be 0 ('recordsWhereValueForResultingColumnShouldBe0')
        recordsWhereValueForResultingColumnShouldBe0 = np.append(recordsWhereValueForResultingColumnShouldBe0, indeciesOfThatColBeing0)
    # deleting indecies where the values of the similar columns are not all 0s (leave only indecies statistically known as the mode or the most repeated indecies -indicating 0 appeared in all the similar columns-) ...
    uniqueIndecies, theirCounts = np.unique(recordsWhereValueForResultingColumnShouldBe0, return_counts=True)
    # if the amount of similar columns is x, an index should have x appearances in 'recordsWhereValueForResultingColumnShouldBe0', as indecies appeared there each time it had a value of 0 in any of the x amount of similar columns
    modes = uniqueIndecies[theirCounts == len(soSimilarBinaryColumns)]
    recordsWhereValueForResultingColumnShouldBe0 = modes
    
    # Assign the value 1 to the resulting column whenever it should be (using 'recordsWhereValueForResultingColumnShouldBe1' as the reference)
    dataframeCopy.loc[recordsWhereValueForResultingColumnShouldBe1, 'resultingColumn'] = 1
    # Assign the value 0 to the resulting column whenever it should be (using 'recordsWhereValueForResultingColumnShouldBe0' as the reference)
    dataframeCopy.loc[recordsWhereValueForResultingColumnShouldBe0, 'resultingColumn'] = 0
    
    #Storing only the resulting column (Series) independently in an object: 'resultingSeries'
    resultingSeries = dataframeCopy['resultingColumn'].copy()

    return resultingSeries
    

#### B. Unifying Process:

##### 1. Columns Telling Whether Exposed to Video Games or Not
> - Student has game console? ('hasPS')
> - Student plays single player games out of school? ('playsSPGames')
> - Student plays co-op games out of school? ('playsCOOP')

In [107]:
df['vidGamesExposure'] = unifySoSimilarBinaryColumns(df, ['playsCOOP', 'hasPS', 'playsSPGames'])

In [108]:
df['vidGamesExposure'].value_counts()

1.0    234107
0.0     40704
Name: vidGamesExposure, dtype: int64

<hr>

##### 2. Columns Telling Whether Using Technology for Entertainment in General or Not
> - Student chats online? ('chatsOnline')
> - Student on social networks? ('onFB')
> - Student using internet for fun? ('internetFun')
> - Student streams music? ('hasMusic')
> - Student uploads content? ('uploads')
> - Student using ICT for fun? ('ICTForFun')

In [109]:
df['usingTechForEntertainment'] = unifySoSimilarBinaryColumns(df, ['chatsOnline', 'onFB', 'internetFun', 'hasMusic', 'uploads', 'ICTForFun'])

In [110]:
df['usingTechForEntertainment'].value_counts()

1.0    286491
Name: usingTechForEntertainment, dtype: int64

<p><center>End of unifying similar columns.</center></p>
<hr>

### Storing Wrangled Data in a File

In [111]:
df.to_csv('pisa2012_Wrangled.csv')

<br>
<br>
<h1><center>End of wrangling.</center></h1>
<hr>